### ISSUE TITLE
how to reduce model.model.shared.num_embeddings？

### ISSUE THREAD

**Issue by masterwang22327** [newfull5/NLLB-200-Distilled-350M-en-ko#1]
i am very glad to see your work on distill nllb.but you just reduce layers size may lead to 350M parameters left.your work is just need en -ko,so you can reduce model.model.shared.num_embeddings to reduce the distilled model parameters agian.so have you did the same work before?i am now working at this.watting for your replay。

**Comment by newfull5** [comment]
Thank you for your interest and kind words
In my work on distilling the model for en-ko translation, I mainly focused on reducing the number of layers. As you mentioned, removing non-essential languages other than Korean could be another approach, but that would require modifying the teacher model’s vocabulary and embedding dimensions as well, which makes knowledge distillation infeasible.

**Comment by masterwang22327** [comment]
thank u for your reply，i have an idea。i use the 3.3B nllb model's sentencepiece.bpe.model to tokenizer en-ko public corpus,get a vocabulary,now the vocabulary only have en-ko token and ther necessary token，and then i modify the M2M100ScaledWordEmbedding(new_vocabulary_size,xxx),transfer the teacher weight by id together ,then distill like your strategy.this may work?i am still work for it.

**Comment by saechanOh** [comment]
Once the vocabulary size changes, the dimension of the one-hot vector after tokenization also has to change, and accordingly, the shape of the embedding layer’s matrix must change as well. I was concerned that this might cause some weight loss, so I haven’t tried this approach.

I can’t say for sure since I haven’t actually done it myself either.

If you get a chance to try it, please share the results!

**Comment by newfull5** [comment]
Once the vocabulary size changes, the dimension of the one-hot vector after tokenization also has to change, and accordingly, the shape of the embedding layer’s matrix must change as well. I was concerned that this might cause some weight loss, so I haven’t tried this approach.

I can’t say for sure since I haven’t actually done it myself either.

If you get a chance to try it, please share the results!

**Comment by newfull5** [comment]
Now that I think about it again, unlike the attention layer, the embedding layer flows linearly, so removing some rows from the embedding matrix doesn't seem to affect the overall embedding vector.
It seems feasible enough.

**Comment by masterwang22327** [comment]
Yes, I have pruned the original vocabulary to keep English and Hebrew. The effect is not reduced, but the number of parameters is reduced from 600M to 400M, and the speed is increased by 80%. This is very exciting. I will continue to do follow-up pruning work, reducing the number of encoder-decoder layers from 12->6, and the number of parameters has been reduced to 200M. Below is my vocabulary screening result.

![Image](https://github.com/user-attachments/assets/e2afde00-c462-48b1-bacf-58df1b7f4614)

**Comment by masterwang22327** [comment]
I also noticed that your evaluation code model.generate uses a single statement. In order to evaluate as quickly as possible, I changed it to bath generate, but the batch generation results will have many irrelevant redundant tokens. Have you encountered this problem?
```
import transformers
import evaluate
from datasets import load_dataset, Dataset
from transformers import AutoTokenizer, AutoModelForSeq2SeqLM,NllbTokenizerFast,M2M100ForConditionalGeneration
import tqdm
import os
import asyncio
import time
os.environ['CUDA_VISIBLE_DEVICES'] = '7'
def read_file(path):
        with open(path, 'r', encoding='utf-8') as file:
            return [line.strip() for line in file.readlines()]
en_dataset=read_file('/workspace/flores200_dataset/dev/eng_Latn.dev')
ko_dataset = read_file('/workspace/flores200_dataset/dev/heb_Hebr.dev')
model = AutoModelForSeq2SeqLM.from_pretrained('/workspace/en_he_winsota/checkpoint-71')
tokenizer = NllbTokenizerFast.from_pretrained('/workspace/en_he_winsota/checkpoint-71')
tokenizer.src_lang='eng_Latn'
tokenizer.tgt_lang='heb_Hebr'
import torch
eval_num=100
batch_encode_num=5
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model.to(device)
reference = ko_dataset[:eval_num]
chrf_metric = evaluate.load("sacrebleu")
txt1='till the public here? No. You want to open the public hearing? Great. All right. The public hearing for Council Bill 161, as amended, is open. May we have the staff report? Jeff Hart. Oh. Champagne...'
txt2="Frankly, it provides clarity that certain applications were not subject to the moratorium on small light parking exemption enacted last August, and those applications may continue to use the small shop parking exemption as existed prior to this proposed text amendment. We had a number of projects that were in the pipeline and the moratorium clearly set forward that they could remain under the rules that under which they had applied. And we just missed that."
tokenizer_res1=tokenizer(txt1, return_tensors='pt',padding='longest',truncation=True,max_length=256).to('cuda')
tokenizer_res2=tokenizer([txt1,txt2,txt1,txt1,txt1], return_tensors='pt',padding='longest',truncation=True,max_length=256).to('cuda')
lan_id=tokenizer.convert_tokens_to_ids('heb_Hebr')
output1 = model.generate(**tokenizer_res1,max_length=256,forced_bos_token_id=lan_id)
res1 = tokenizer.batch_decode(output1)
output2 =model.generate(**tokenizer_res2,max_length=256,forced_bos_token_id=lan_id)
res2 = tokenizer.batch_decode(output2)
prediction = []
eval_rank=0
while eval_rank<eval_num:
    task=[]
    sentences=en_dataset[eval_rank:eval_rank+batch_encode_num]
    
    tokenizer_res=tokenizer(sentences, return_tensors='pt',
                        padding='longest',truncation=True,
                        max_length=256).to('cuda')
    output = model.generate(**tokenizer_res,
                            forced_bos_token_id=tokenizer.convert_tokens_to_ids('heb_Hebr'),
                            max_length=256)
    output = tokenizer.batch_decode(output )
    prediction.extend(output)
    eval_rank+=batch_encode_num
chrf_metric.add_batch(
    predictions=prediction,
    references=reference
)
print(chrf_metric.compute())
```
